source/ed/build_helper.cpp

```diff
@@ -100,16 +100,31 @@ VJ::VJ(builder & b, const std::string &line_name, const std::string &validity_pa
     jp->route = route;
 
     //add physical mode
-    if (! b.data->pt_data->physical_modes.empty()) {
-        auto mode = b.data->pt_data->physical_modes.front();
-        jp->physical_mode = mode;
+    if (!physical_mode.empty()){
+        auto it = b.data->pt_data->physical_modes_map.find(physical_mode);
+        if (it != b.data->pt_data->physical_modes_map.end()){
+            jp->physical_mode = it->second;
+        }
     }
-    else {
-        jp->physical_mode = new navitia::type::PhysicalMode();
-        jp->physical_mode->idx = b.data->pt_data->physical_modes.size();
-        jp->physical_mode->uri = "physical_mode:0";
-        b.data->pt_data->physical_modes.push_back(jp->physical_mode);
+    if(!jp->physical_mode){
+        if(!physical_mode.empty()){
```

Comment:
you can factorize some code here no ?

if (physical_mode.empty() && b.data->pt_data->physical_modes.size()) { jp->physical_mode = b.data->pt_data->physical_modes.front();}
else {
const auto name = physicla_mode.empty() : "physical_mode:0" : physical_mode;
                jp->physical_mode = new navitia::type::PhysicalMode();
                jp->physical_mode->idx = b.data->pt_data->physical_modes.size();
                jp->physical_mode->uri = name;
                jp->physical_mode->name = name;
                b.data->pt_data->physical_modes.push_back(jp->physical_mode);
}